table.go

```diff
@@ -47,31 +57,133 @@ type RoutingTable struct {
 	maxLatency time.Duration
 
 	// kBuckets define all the fingers to other nodes.
-	Buckets    []*Bucket
+	Buckets    []*bucket
 	bucketsize int
 
 	cplRefreshLk   sync.RWMutex
 	cplRefreshedAt map[uint]time.Time
 
+	// replacement candidates for a Cpl
+	cplReplacementCache *cplReplacementCache
+
 	// notification functions
 	PeerRemoved func(peer.ID)
 	PeerAdded   func(peer.ID)
+
+	// function to determine the validity of a peer for RT membership
+	PeerValidationFnc PeerValidationFnc
+
+	// timeout for a single call to the peer validation function
+	peerValidationTimeout time.Duration
+	// interval between two runs of the table cleanup routine
+	tableCleanupInterval time.Duration
+	// function to select peers that need to be validated
+	peersForValidationFnc PeerSelectionFnc
 }
 
 // NewRoutingTable creates a new routing table with a given bucketsize, local ID, and latency tolerance.
-func NewRoutingTable(bucketsize int, localID ID, latency time.Duration, m peerstore.Metrics) *RoutingTable {
+// Passing a nil PeerValidationFnc disables periodic table cleanup.
+func NewRoutingTable(ctx context.Context, bucketsize int, localID ID, latency time.Duration, m peerstore.Metrics,
```

Comment:
Just curious; why not make the `peerValidationFnc` one of the `Option`s? We could potentially convert the other args (latency, etc) into `Option`s as well, as long as we're changing the signature.

Reply:
Using contexts to stop services is really an anti-pattern (yeah, I know we use it everywhere...). We should provide a close function.

Otherwise, the routing table could stop _before_ the DHT (context cancellation isn't ordered).

Reply:
@Stebalien Could you elaborate a bit on this ? Why is it an anti-pattern ? Also, could you direct me to some code which uses a Close function like you mention ?

Reply:
@yusefnapora Valid point. I've made the change.

Reply:
Context cancellation is designed for requests: https://golang.org/pkg/context/.

* Previous answer: https://github.com/ipfs/go-bitswap/pull/248#discussion_r377975239
* Lots of discussion here: https://github.com/ipfs/go-ipfs/issues/5810

Basically, if we use contexts, everything will shutdown at the same time in a random order without taking dependencies into account. If we use explicit _close_ methods, we can close a service and then have it close the sub-components it owns, etc.

We've had problems in the past where we'd, e.g., close the datastore before we're done using it.

Reply:
@Stebalien  Thanks for this. I've made the change. Also dug through the `goprocess` library for this & now have a pretty good idea of how it works internally.

Reply:
Personally speaking I don't think using context this way is bad. `goprocess` is the antipattern, not context + cancellation.